contracts/RockPaperScissors.sol

```diff
@@ -0,0 +1,267 @@
+        require(game.playerOneMoveHash != bytes32(0) && game.playerTwoMoveHash != bytes32(0), GAME_IN_PROGRESS_MSG);
+
+        bytes32 expectedMoveToken = hashPlayerMove(gameToken, msg.sender, secret, playerMove);
+
+        if (msg.sender == game.playerOne) {
+            require(game.playerOneMoveHash == expectedMoveToken, BAD_MATCH_MSG);
+            game.playerOneMove = playerMove;
+        }
+
+        if (msg.sender == game.playerTwo) {
+            require(game.playerTwoMoveHash == expectedMoveToken, BAD_MATCH_MSG);
+            game.playerTwoMove = playerMove;
+        }
+
+        if (game.playerOneMove != 0 && game.playerTwoMove != 0) {
+            emit AllMovesRevealed(gameToken);
+        }
+    }
+
+    function determineWinner(bytes32 gameToken) external onlyOwner whenNotPaused {
+        Game storage game = games[gameToken];
+        require(game.stake != 0, GAME_NOT_FOUND_MSG);
+
+        if(game.playerOneMove == game.playerTwoMove){
+            resetGame(game);
+            emit Draw(gameToken);
+            return;
+        }
+
+        address winner;
+
+        if (isWinningMove(game.playerOneMove, game.playerTwoMove)) {
+            winner = game.playerOne;
+        } else {
+            winner = game.playerTwo;
+        }
+
+        uint winnings = game.stake * 2;
+        resetGame(game);
+        emit WinnerRewarded(gameToken, winner, winnings);
+
+        balances[winner] = balances[winner].add(winnings);
+    }
+
+    function isWinningMove(uint8 playerOneMove, uint8 playerTwoMove) internal pure returns (bool) {
+
+        if (playerOneMove == ROCK && playerTwoMove == SCISSORS) {
+            return true;
+        }
+        if (playerOneMove == PAPER && playerTwoMove == ROCK) {
+            return true;
+        }
+        if (playerOneMove == SCISSORS && playerTwoMove == PAPER) {
+            return true;
+        }
+
+        return false;
+    }
+
+    function withdraw(uint amount) external whenNotPaused returns(bool success) {
+        uint256 withdrawerBalance = balances[msg.sender];
+        require(amount > 0, "The value must be greater than 0");
+        require(withdrawerBalance >= amount, "There are insufficient funds");
+
+        emit WithDraw(msg.sender, amount);
+
+        balances[msg.sender] = SafeMath.sub(withdrawerBalance, amount);
+        (success, ) = msg.sender.call{value: amount}("");
+        require(success, "Transfer failed");
+    }
+
+    function withDrawFromGame(bytes32 gameToken) whenNotPaused external whenNotPaused {
+        Game storage game = games[gameToken];
+        uint256 stake = game.stake;
+        require(stake != 0, GAME_NOT_FOUND_MSG);
+        require(msg.sender == game.playerOne || msg.sender == game.playerTwo, INVALID_PLAYER_MSG);
+
+        if (msg.sender == game.playerOne) {
+            require(game.playerTwoMoveHash == bytes32(0), GAME_IN_PROGRESS_MSG);
+            require(game.playerOneMoveHash != bytes32(0), NO_STAKE_TO_WITHDRAW_MSG);
+        }
```

Comment:
Can player one front-run player2 and withdraw if they see that player2 is going to win with a known hash?

Reply:
Oh. I had not considered that!

Reply:
I have been considering a new strategy.

The player who makes the initial wager must place a security deposit which exceeds the winning total. This is in addition to the stake. They must reveal their move in the final step even if they know they have lost or they forfeit both their stake and the security deposit.

Reply:
Mmh, that's asking more of them. Perhaps they should be rewarded for this extra risk and extra lock-up of funds.
Or you make the forfeit deadline short enough so that the winner by forfeit does not want much longer than if the loser had revealed.

Reply:
So thinking about it again without a security deposit. If I introduce a deadline and the other player does not reveal their 'losing' move by the deadline expiry then we assume they have lost and the other player becomes the winner by default.

The player setting up the game will be the one to to have to pay the cost of calling initialise() so perhaps should be compensated in some way for this.

Reply:
> so perhaps should be compensated in some way for this.

Let's see how much it really costs. i.e., how many storage slots are touched. And how does it compare to the other function calls.